### ISSUE TITLE
A Neat XOR Trick

### ISSUE THREAD

**Issue by luantranminh** [luantranminh/til#147]
https://www.mattkeeter.com/blog/2022-12-10-xor/
Problem summary, you're given an input string, e.g.
`nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg`
Within this string, you want to find the first W-character window in which every character is unique.

For this example, with a 4-character window, this occurs here:

`nznrnf [rfnt] jfmvfwmzdfjlvtqnbhcprsg`
You could build a HashSet and check its size:
```rust
fn run(s: &[char], window_size: usize) -> usize {
    for i in 0..s.len() - window_size {
        let mut set = HashSet::new();
        for j in 0..window_size {
            set.insert(s[i + j]);
        }
        if set.len() == window_size {
            return i + window_size;
        }
    }
    panic!("No unique window found");
}
```
Insertion into the HashSet is O(1), so our running time is down to O(N * W).

We can consider each character as a bitmask. The **OR (|)** of their bitmasks will have W bits set if and only if those characters are unique.
```
a | b | c = 11100000000000000000000000000000
a | b | b = 11000000000000000000000000000000
```
If we have a set of W characters, the [xor](https://en.wikipedia.org/wiki/Exclusive_or) of their bitmasks will have W bits set if and only if those characters are unique.
```
a ^ b ^ c = 11100000000000000000000000000000
a ^ b ^ b = 10000000000000000000000000000000
```
Using **xor** is more powerful than **or**, because of one particular property:
```
a ^ b ^ a = b
```
Applying xor twice will turn a bit on, then back off.

In other words, we can maintain the bitset for a sliding window in a single u32, turning bits on as they enter the window and off as they leave.

This eliminates the inner loop from our previous function:

```rust
fn run(s: &[char], window_size: usize) -> usize {
    let mut set = 0u32;
    for i in 0..s.len() {
        // Turn on bits as they enter the window
        set ^= 1 << (s[i] as u32 - 'a' as u32);

        // Turn off bits as they leave the window
        if i >= window_size {
            set ^= 1 << (s[i - window_size] as u32 - 'a' as u32);
        }

        // Check the current window and see if we're done
        if set.count_ones() as usize == window_size {
            return i + 1;
        }
    }
    panic!("No unique window found");
}
```
